p/api/fever.php

```diff
@@ -0,0 +1,894 @@
+			FreshRSS_Context::$user_conf = null;
+			return;
+		}
+
+		$apiKey = $_POST["api_key"];
+		$usersDir = join_path(DATA_PATH, 'users');
+
+		foreach (glob($usersDir . '/*', GLOB_ONLYDIR) as $username) {
+			$username = str_replace($usersDir . '/', '', $username);
+			if ($username == '_') {
+				continue;
+			}
+
+			$config = get_user_configuration($username);
+			if ($config->apiPasswordHash != '' && password_verify($apiKey, $config->apiPasswordHash)) {
+				Minz_Session::_param('currentUser', $username);
+				FreshRSS_Context::$user_conf = $config;
+				return;
+			}
+		}
+	}
+
+	/**
+	 * @return bool
+	 */
+	public function isXmlRequested()
+	{
+		return $this->xml;
+	}
+
+	/**
+	 * @return bool
+	 */
+	public function isAuthenticatedApiUser()
+	{
+		if (FreshRSS_Context::$user_conf !== null) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * @return FeverAPI_FeedDAO
+	 */
+	protected function getDaoForFeeds()
+	{
+		return new FeverAPI_FeedDAO();
+	}
+
+	/**
+	 * @return FreshRSS_CategoryDAO
+	 */
+	protected function getDaoForCategories()
+	{
+		return new FreshRSS_CategoryDAO();
+	}
+
+	/**
+	 * @return FeverAPI_EntryDAO
+	 */
+	protected function getDaoForEntries()
+	{
+		return new FeverAPI_EntryDAO();
+	}
+
+	/**
+	 * this does all the processing, since the fever api does not have a specific variable that specifies the operation
+	 */
+	public function process()
+	{
+		$response_arr = array();
+
+		if (!$this->isAuthenticatedApiUser()) {
+			throw new \Exception('No user given or user is not allowed to access API');
+		}
+
+		if (isset($_REQUEST["groups"])) {
+			$response_arr["groups"] = $this->getGroups();
+			$response_arr["feeds_groups"] = $this->getFeedsGroup();
+		}
+
+		if (isset($_REQUEST["feeds"])) {
+			$response_arr["feeds"] = $this->getFeeds();
+			$response_arr["feeds_groups"] = $this->getFeedsGroup();
+		}
+
+		if (isset($_REQUEST["favicons"])) {
+			$response_arr["favicons"] = $this->getFavicons();
+		}
+
+		if (isset($_REQUEST["items"])) {
+			$response_arr["total_items"] = $this->getTotalItems();
+			$response_arr["items"] = $this->getItems();
+		}
+
+		if (isset($_REQUEST["links"])) {
+			$response_arr["links"] = $this->getLinks();
+		}
+
+		if (isset($_REQUEST["unread_item_ids"])) {
+			$response_arr["unread_item_ids"] = $this->getUnreadItemIds();
+		}
+
+		if (isset($_REQUEST["saved_item_ids"])) {
+			$response_arr["saved_item_ids"] = $this->getSavedItemIds();
+		}
+
+		if (isset($_REQUEST["mark"], $_REQUEST["as"], $_REQUEST["id"]) && is_numeric($_REQUEST["id"])) {
+			$method_name = "set" . ucfirst($_REQUEST["mark"]) . "As" . ucfirst($_REQUEST["as"]);
+			if (method_exists($this, $method_name)) {
+				$id = intval($_REQUEST["id"]);
+				switch (strtolower($_REQUEST["mark"])) {
+					case 'item':
+						$this->{$method_name}($id);
+						break;
+					case 'feed':
+					case 'group':
+						$before = (isset($_REQUEST["before"])) ? $_REQUEST["before"] : null;
+						if ($before > pow(10, 10)) {
```

Comment:
That is a constant, right? If yes, let's avoid calling an expensive function.

Reply:
I will remove this, as we have to handle the before timestamp completely different